FreeRTOS-Plus/Source/Application-Protocols/network_transport/freertos_plus_tcp/using_wolfSSL/using_wolfSSL.c

```diff
@@ -0,0 +1,409 @@
+    }
+    else
+    {
+        /* do nothing */
+    }
+    return (int)sent;
+}
+
+/*-----------------------------------------------------------*/
+static TlsTransportStatus_t initTLS(void)
+{
+    /* initialize wolfSSL */
+    wolfSSL_Init();
+
+#ifdef DEBUG_WOLFSSL
+    wolfSSL_Debugging_ON();
+#endif
+
+    return TLS_TRANSPORT_SUCCESS;
+}
+
+/*-----------------------------------------------------------*/
+
+static TlsTransportStatus_t tlsSetup( NetworkContext_t * pNetworkContext,
+                                      const char * pHostName,
+                                      const NetworkCredentials_t * pNetworkCredentials )
+{
+    TlsTransportStatus_t returnStatus = TLS_TRANSPORT_SUCCESS;
+    int iResult = SSL_SUCCESS;
+    /*char errString[80];*/
+
+    configASSERT( pNetworkContext != NULL );
+    configASSERT( pHostName != NULL );
+    configASSERT( pNetworkCredentials != NULL );
+    configASSERT( pNetworkCredentials->pRootCa != NULL );
+    configASSERT( pNetworkContext->tcpSocket != NULL );
+    
+    if (pNetworkContext->sslContext.ctx == NULL)
+    {
+        /* Attempt to create a context that uses the TLS 1.3 or 1.2 client protocol. */
+        pNetworkContext->sslContext.ctx = wolfSSL_CTX_new(wolfSSLv23_client_method_ex(NULL));
+    }
+
+    if (pNetworkContext->sslContext.ctx == NULL)
+    {
+        LogError(("Failed to create a wolfSSL_CTX"));
+        returnStatus = TLS_TRANSPORT_CONNECT_FAILURE;
+        return returnStatus;
+    }
+
+    /* attempt to load ca cert file, client cert file and client private key file */
+    iResult = wolfSSL_CTX_load_verify_locations(pNetworkContext->sslContext.ctx,
+        (const char*)(pNetworkCredentials->pRootCa), NULL);
+
+    if (iResult == SSL_SUCCESS)
+    {
+        iResult = wolfSSL_CTX_use_certificate_file(pNetworkContext->sslContext.ctx,
+            (const char*)(pNetworkCredentials->pClientCert), SSL_FILETYPE_PEM);
+    }
+    else
+    {
+        LogError(("Failed to load ca-certificate file"));
+        returnStatus = TLS_TRANSPORT_INVALID_CREDENTIALS;
+    }
+ 
+    if (iResult == SSL_SUCCESS)
+    {
+        iResult = wolfSSL_CTX_use_PrivateKey_file(pNetworkContext->sslContext.ctx,
+            (const char*)(pNetworkCredentials->pPrivateKey), SSL_FILETYPE_PEM);
+    }
+    else
+    {
+        LogError(("Failed to load client-certificate file"));
+        returnStatus = TLS_TRANSPORT_INVALID_CREDENTIALS;
+    }
+
+    if (iResult != SSL_SUCCESS)
+    {
+        LogError(("Failed to load client-private-key file"));
+        returnStatus = TLS_TRANSPORT_INVALID_CREDENTIALS;
+    }
+ 
+    /* on success of file loading, attempt to create a ssl object */
+    if (returnStatus == TLS_TRANSPORT_SUCCESS)
+    {
+        /* create a ssl object */
+        pNetworkContext->sslContext.ssl = wolfSSL_new(pNetworkContext->sslContext.ctx);
+        Socket_t xSocket = pNetworkContext->tcpSocket;
+
+        if (pNetworkContext->sslContext.ssl != NULL)
```

Comment:
If this fails, should an error be returned?

Reply:
Fixed to return an error